File: samples/hellofs_cs/hellofs.cs
//----------------------------------------------------------------------------
// Copyright 2008-2015 Joe Lowe
//
// Permission is granted to any person obtaining a copy of this Software,
// to deal in the Software without restriction, including the rights to use,
// copy, modify, merge, publish, distribute, sublicense, and sell copies of
// the Software.
//
// The above copyright and permission notice must be left intact in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS WITHOUT WARRANTY.
//----------------------------------------------------------------------------
// file name:  hellofs.cs
// created:    2008.04.25
//    Hello World file system sample application.
//----------------------------------------------------------------------------
using System;
using System.Text;
using System.IO;

class Volume: Pfm.FormatterDispatch
{
   public Pfm.Marshaller marshaller;
   public long folderOpenId = 0;
   public long fileOpenId = 0;
   static string helloFileName = "readme.txt";
   static string helloFileData = "Hello World.\r\n";
   static int helloDataSize = helloFileData.Length;
   static long helloRootId = 2;
   static long helloFileId = 3;

   public void Dispose()
   {
   }

   public void Open( Pfm.MarshallerOpenOp op)
   {
      string[] nameParts = op.NameParts();
      int createFileType = op.CreateFileType();
      long newExistingOpenId = op.NewExistingOpenId();
      int perr = 0;
      bool existed = false;
      Pfm.OpenAttribs openAttribs = new Pfm.OpenAttribs();
      long parentFileId = 0;
      string endName = null;

      if (nameParts.Length == 0)
      {
         if (folderOpenId == 0)
         {
            folderOpenId = newExistingOpenId;
         }
         existed = true;
         openAttribs.openId = folderOpenId;
         openAttribs.openSequence = 1;
         openAttribs.accessLevel = Pfm.accessLevelReadData;
         openAttribs.attribs.fileType = Pfm.fileTypeFolder;
         openAttribs.attribs.fileId = helloRootId;
      }
      else if(nameParts.Length == 1)
      {
         if(nameParts[0].ToLowerInvariant() != helloFileName.ToLowerInvariant())
         {
            perr = Pfm.errorNotFound;
         }
         else
         {
            if (fileOpenId == 0)
            {
               fileOpenId = newExistingOpenId;
            }
            existed = true;
            openAttribs.openId = fileOpenId;
            openAttribs.openSequence = 1;
            openAttribs.accessLevel = Pfm.accessLevelReadData;
            openAttribs.attribs.fileType = Pfm.fileTypeFile;
            openAttribs.attribs.fileId = helloFileId;
            openAttribs.attribs.fileSize = helloFileData.Length;
            endName = helloFileName;
         }
      }
      else
      {
         perr = Pfm.errorParentNotFound;
      }
      if(perr == Pfm.errorNotFound && createFileType != 0)
      {
         perr = Pfm.errorAccessDenied;
      }

      op.Complete( perr, existed, openAttribs, parentFileId, endName, 0, null, 0, null);
   }

   public void Replace( Pfm.MarshallerReplaceOp op)
   {
      op.Complete( Pfm.errorAccessDenied, null/*openAttribs*/, null);
   }

   public void Move( Pfm.MarshallerMoveOp op)
   {
      op.Complete( Pfm.errorAccessDenied, false/*existed*/, null/*openAttribs*/, 0/*parentFileId*/, null/*endName*/, 0, null, 0, null);
   }

   public void MoveReplace( Pfm.MarshallerMoveReplaceOp op)
   {
      op.Complete( Pfm.errorAccessDenied);
   }

   public void Delete( Pfm.MarshallerDeleteOp op)
   {
      op.Complete( Pfm.errorAccessDenied);
   }

   public void Close( Pfm.MarshallerCloseOp op)
   {
      op.Complete( Pfm.errorSuccess);
   }

   public void FlushFile( Pfm.MarshallerFlushFileOp op)
   {
      long openId = op.OpenId();
      int fileFlags = op.FileFlags();
      int color = op.Color();
      long createTime = op.CreateTime();
      long writeTime = op.WriteTime();
      int perr = 0;
      Pfm.OpenAttribs openAttribs = new Pfm.OpenAttribs();

      if (fileFlags != Pfm.fileFlagsInvalid ||
         color != Pfm.colorInvalid ||
         createTime != Pfm.timeInvalid ||
         writeTime != Pfm.timeInvalid)
      {
         perr = Pfm.errorAccessDenied;
      }
      else if (openId == folderOpenId)
      {
         openAttribs.openId = folderOpenId;
         openAttribs.openSequence = 1;
         openAttribs.accessLevel = Pfm.accessLevelReadData;
         openAttribs.attribs.fileType = Pfm.fileTypeFolder;
         openAttribs.attribs.fileId = helloRootId;
      }
      else if (openId == fileOpenId)
      {
         openAttribs.openId = fileOpenId;
         openAttribs.openSequence = 1;
         openAttribs.accessLevel = Pfm.accessLevelReadData;
         openAttribs.attribs.fileType = Pfm.fileTypeFile;
         openAttribs.attribs.fileId = helloFileId;
         openAttribs.attribs.fileSize = helloFileData.Length;
      }
      else
      {
         perr = Pfm.errorNotFound;
      }

      op.Complete( perr, openAttribs, null);
   }

   public void List( Pfm.MarshallerListOp op)
   {
      long openId = op.OpenId();
      int perr = 0;
      Pfm.Attribs attribs = new Pfm.Attribs();

      if (openId != folderOpenId)
      {
         perr = Pfm.errorAccessDenied;
      }
      else
      {
         attribs.fileType = Pfm.fileTypeFile;
         attribs.fileId = helloFileId;
         attribs.fileSize = helloFileData.Length;
         op.Add( attribs, helloFileName);
      }

      op.Complete( perr, true/*noMore*/);
   }

   public void ListEnd( Pfm.MarshallerListEndOp op)
   {
      op.Complete( Pfm.errorSuccess);
   }

   public void Read( Pfm.MarshallerReadOp op)
   {
      long openId = op.OpenId();
      long fileOffset = op.FileOffset();
      byte[] data = op.Data();
      int requestedSize = op.RequestedSize();
      int perr = 0;
      int actualSize = 0;
      int i;

      if (openId != fileOpenId)
      {
         perr = Pfm.errorAccessDenied;
      }
      else
      {
         actualSize = requestedSize;
         if (fileOffset > helloDataSize)
         {
            actualSize = 0;
         }
         else if (fileOffset+requestedSize > helloDataSize)
         {
            actualSize = (int)(helloDataSize-fileOffset);
         }
         if (actualSize != 0)
         {
            for (i = 0; i < actualSize; i++)
            {
               data[i] = (byte)(helloFileData[(int)fileOffset+i]);
            }
         }
      }

      op.Complete( perr, actualSize);
   }

   public void Write( Pfm.MarshallerWriteOp op)
   {
      op.Complete( Pfm.errorAccessDenied, 0/*actualSize*/);
   }

   public void SetSize( Pfm.MarshallerSetSizeOp op)
   {
      op.Complete( Pfm.errorAccessDenied);
   }

   public void Capacity( Pfm.MarshallerCapacityOp op)
   {
      op.Complete( Pfm.errorSuccess, helloFileData.Length/*totalCapacity*/, 0/*availableCapacity*/);
   }

   public void FlushMedia( Pfm.MarshallerFlushMediaOp op)
   {
      op.Complete( Pfm.errorSuccess, -1/*flushTimeoutMsecs*/);
   }

   public void Control( Pfm.MarshallerControlOp op)
   {
      op.Complete( Pfm.errorInvalid, 0/*outputSize*/);
   }

   public void MediaInfo( Pfm.MarshallerMediaInfoOp op)
   {
      Pfm.MediaInfo mediaInfo = new Pfm.MediaInfo();

      op.Complete( Pfm.errorSuccess, mediaInfo, "hellofs"/*mediaLabel*/);
   }

   public void Access( Pfm.MarshallerAccessOp op)
   {
      long openId = op.OpenId();
      int perr = 0;
      Pfm.OpenAttribs openAttribs = new Pfm.OpenAttribs();

      if (openId == folderOpenId)
      {
         openAttribs.openId = folderOpenId;
         openAttribs.openSequence = 1;
         openAttribs.accessLevel = Pfm.accessLevelReadData;
         openAttribs.attribs.fileType = Pfm.fileTypeFolder;
         openAttribs.attribs.fileId = helloRootId;
      }
      else if(openId == fileOpenId)
      {
         openAttribs.openId = fileOpenId;
         openAttribs.openSequence = 1;
         openAttribs.accessLevel = Pfm.accessLevelReadData;
         openAttribs.attribs.fileType = Pfm.fileTypeFile;
         openAttribs.attribs.fileId = helloFileId;
         openAttribs.attribs.fileSize = helloFileData.Length;
      }
      else
      {
         perr = Pfm.errorNotFound;
      }

      op.Complete( perr, openAttribs, null);
   }

   public void ReadXattr( Pfm.MarshallerReadXattrOp op)
   {
      op.Complete( Pfm.errorNotFound, 0/*xattrSize*/, 0/*transferredSize*/);
   }

   public void WriteXattr( Pfm.MarshallerWriteXattrOp op)
   {
      op.Complete( Pfm.errorAccessDenied, 0/*transferredSize*/);
   }
}

public class Main_
{

static void Main( string[] args)
{
   int err = 0;
   Pfm.Api pfm = null;
   Pfm.Mount mount = null;
   Pfm.MountCreateParams mcp = new Pfm.MountCreateParams();
   Pfm.MarshallerServeParams msp = new Pfm.MarshallerServeParams();
   Volume volume = new Volume();
   IntPtr invalidFd = new IntPtr( -1);

   msp.dispatch = volume;
   msp.formatterName = "hellofs";

   if (args.Length != 1)
   {
      Console.Write(
         "Sample file system application.\n" +
         "syntax: hellofs <mount name>\n");
      err = -1;
   }
   if (err == 0)
   {
      mcp.mountSourceName = args[0];
      err = Pfm.ApiFactory( out pfm);
      if(err != 0)
      {
         Console.Write( "ERROR: {0} Unable to open PFM API.\n", err);
      }
   }
   if (err == 0)
   {
      err = Pfm.MarshallerFactory( out volume.marshaller);
      if(err != 0)
      {
         Console.Write( "ERROR: {0} Unable to create marshaller.\n", err);
      }
   }
   if (err == 0)
   {
         // Communication between the driver and file system is done
         // over a pair of simple pipes. Application is responsible
         // for creating the pipes.
      err = Pfm.SystemCreatePipe( out msp.toFormatterRead, out mcp.toFormatterWrite);
      if(err == 0)
      {
         err = Pfm.SystemCreatePipe( out mcp.fromFormatterRead, out msp.fromFormatterWrite);
      }
   }
   if (err == 0)
   {
         // Various mount options are available through mountFlags,
         // visibleProcessId, and ownerSid.
      err = pfm.MountCreate( mcp, out mount);
      if (err != 0)
      {
         Console.Write( "ERROR: {0} Unable to create mount.\n", err);
      }
   }

      // Close driver end pipe handles now. Driver has duplicated what
      // it needs. If these handles are not closed then pipes will not
      // break if driver disconnects, leaving us stuck in the
      // marshaller.
   Pfm.SystemCloseFd( mcp.toFormatterWrite);
   Pfm.SystemCloseFd( mcp.fromFormatterRead);

   if (err == 0)
   {
         // If tracemon is installed and running then diagnostic
         // messsages can be viewed in the "hellofs" channel.
      volume.marshaller.SetTrace( mount.GetMountEndName());
         // Also send diagnostic messages to stdout. This can slow
         // things down quite a bit.
      IntPtr stdoutFd = invalidFd;
      Pfm.SystemGetStdout( out stdoutFd);
      volume.marshaller.SetStatus( stdoutFd);

         // The marshaller uses alertable I/O, so process can be
         // exited via ctrl+c.
      Console.Write( "Press CTRL+C to exit.\n");
         // The marshaller serve function will return at unmount or
         // if driver disconnects.
      volume.marshaller.ServeDispatch( msp);
   }

   Pfm.SystemCloseFd( msp.toFormatterRead);
   Pfm.SystemCloseFd( msp.fromFormatterWrite);
   if (pfm != null)
   {
      pfm.Dispose();
   }
   if (mount != null)
   {
      mount.Dispose();
   }
   if (volume.marshaller != null)
   {
      volume.marshaller.Dispose();
   }
}

}

File: samples/hellofs_cpp/hellofs.cpp
//----------------------------------------------------------------------------
// Copyright 2006-2015 Joe Lowe
//
// Permission is granted to any person obtaining a copy of this Software,
// to deal in the Software without restriction, including the rights to use,
// copy, modify, merge, publish, distribute, sublicense, and sell copies of
// the Software.
//
// The above copyright and permission notice must be left intact in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS WITHOUT WARRANTY.
//----------------------------------------------------------------------------
// file name:  hellofs.cpp
// created:    2006.10.06
//    Hello world file system sample application.
//----------------------------------------------------------------------------

#include "../common/portability.c"

#include "pfmapi.h"
#define CCALL PT_CCALL

static PfmOpenAttribs zeroOpenAttribs = {};
static PfmAttribs zeroAttribs = {};
static PfmMediaInfo zeroMediaInfo = {};

static const wchar_t helloFileName[] = L"readme.txt";
static const char helloData[] = "Hello world.\r\n";
static const size_t helloDataSize = sizeof( helloData)-sizeof( helloData[0]);
static const int64_t helloRootId = 2;
static const int64_t helloFileId = 3;

struct Volume: PfmFormatterDispatch
{
   PfmMarshaller* marshaller;
   int64_t folderOpenId;
   int64_t fileOpenId;

   Volume(void);
   ~Volume(void);

      // PfmFormatterDispatch
   void CCALL Open( PfmMarshallerOpenOp* op, void* formatterUse);
   void CCALL Replace( PfmMarshallerReplaceOp* op, void* formatterUse);
   void CCALL Move( PfmMarshallerMoveOp* op, void* formatterUse);
   void CCALL MoveReplace( PfmMarshallerMoveReplaceOp* op, void* formatterUse);
   void CCALL Delete( PfmMarshallerDeleteOp* op, void* formatterUse);
   void CCALL Close( PfmMarshallerCloseOp* op, void* formatterUse);
   void CCALL FlushFile( PfmMarshallerFlushFileOp* op, void* formatterUse);
   void CCALL List( PfmMarshallerListOp* op, void* formatterUse);
   void CCALL ListEnd( PfmMarshallerListEndOp* op, void* formatterUse);
   void CCALL Read( PfmMarshallerReadOp* op, void* formatterUse);
   void CCALL Write( PfmMarshallerWriteOp* op, void* formatterUse);
   void CCALL SetSize( PfmMarshallerSetSizeOp* op, void* formatterUse);
   void CCALL Capacity( PfmMarshallerCapacityOp* op, void* formatterUse);
   void CCALL FlushMedia( PfmMarshallerFlushMediaOp* op, void* formatterUse);
   void CCALL Control( PfmMarshallerControlOp* op, void* formatterUse);
   void CCALL MediaInfo( PfmMarshallerMediaInfoOp* op, void* formatterUse);
   void CCALL Access( PfmMarshallerAccessOp* op, void* formatterUse);
   void CCALL ReadXattr( PfmMarshallerReadXattrOp* op, void* formatterUse);
   void CCALL WriteXattr( PfmMarshallerWriteXattrOp* op, void* formatterUse);
};

void CCALL Volume::Open( PfmMarshallerOpenOp* op, void* formatterUse)
{
   const PfmNamePart* nameParts = op->NameParts();
   size_t namePartCount = op->NamePartCount();
   int8_t createFileType = op->CreateFileType();
   int64_t newExistingOpenId = op->NewExistingOpenId();
   int perr = 0;
   bool existed = false;
   PfmOpenAttribs openAttribs = zeroOpenAttribs;
   int64_t parentFileId = 0;
   const wchar_t* endName = 0;

   if (!namePartCount)
   {
      if (!folderOpenId)
      {
         folderOpenId = newExistingOpenId;
      }
      existed = true;
      openAttribs.openId = folderOpenId;
      openAttribs.openSequence = 1;
      openAttribs.accessLevel = pfmAccessLevelReadData;
      openAttribs.attribs.fileType = pfmFileTypeFolder;
      openAttribs.attribs.fileId = helloRootId;
   }
   else if (namePartCount == 1)
   {
      if (sswcmpf(nameParts[0].name,helloFileName) != 0)
      {
         perr = pfmErrorNotFound;
      }
      else
      {
         if (!fileOpenId)
         {
            fileOpenId = newExistingOpenId;
         }
         existed = true;
         openAttribs.openId = fileOpenId;
         openAttribs.openSequence = 1;
         openAttribs.accessLevel = pfmAccessLevelReadData;
         openAttribs.attribs.fileType = pfmFileTypeFile;
         openAttribs.attribs.fileId = helloFileId;
         openAttribs.attribs.fileSize = helloDataSize;
         endName = helloFileName;
      }
   }
   else
   {
      perr = pfmErrorParentNotFound;
   }
   if (perr == pfmErrorNotFound && createFileType != 0)
   {
      perr = pfmErrorAccessDenied;
   }

   op->Complete( perr, existed, &openAttribs, parentFileId, endName, 0, 0, 0, 0);
}

void CCALL Volume::Replace( PfmMarshallerReplaceOp* op, void* formatterUse)
{
   op->Complete( pfmErrorAccessDenied, 0/*openAttribs*/, 0);
}

void CCALL Volume::Move( PfmMarshallerMoveOp* op, void* formatterUse)
{
   op->Complete( pfmErrorAccessDenied, 0/*existed*/, 0/*openAttribs*/, 0/*parentFileId*/, 0/*endName*/, 0, 0, 0, 0);
}

void CCALL Volume::MoveReplace( PfmMarshallerMoveReplaceOp* op, void* formatterUse)
{
   op->Complete( pfmErrorAccessDenied);
}

void CCALL Volume::Delete( PfmMarshallerDeleteOp* op, void* formatterUse)
{
   op->Complete( pfmErrorAccessDenied);
}

void CCALL Volume::Close( PfmMarshallerCloseOp* op, void* formatterUse)
{
   op->Complete( pfmErrorSuccess);
}

void CCALL Volume::FlushFile( PfmMarshallerFlushFileOp* op, void* formatterUse)
{
   int64_t openId = op->OpenId();
   uint8_t fileFlags = op->FileFlags();
   uint8_t color = op->FileFlags();
   int64_t createTime = op->CreateTime();
   int64_t writeTime = op->WriteTime();
   int perr = 0;
   PfmOpenAttribs openAttribs = zeroOpenAttribs;

   if (fileFlags != pfmFileFlagsInvalid ||
      color != pfmColorInvalid ||
      createTime != pfmTimeInvalid ||
      writeTime != pfmTimeInvalid)
   {
      perr = pfmErrorAccessDenied;
   }
   else if (openId == folderOpenId)
   {
      openAttribs.openId = folderOpenId;
      openAttribs.openSequence = 1;
      openAttribs.accessLevel = pfmAccessLevelReadData;
      openAttribs.attribs.fileType = pfmFileTypeFolder;
      openAttribs.attribs.fileId = helloRootId;
   }
   else if (openId == fileOpenId)
   {
      openAttribs.openId = fileOpenId;
      openAttribs.openSequence = 1;
      openAttribs.accessLevel = pfmAccessLevelReadData;
      openAttribs.attribs.fileType = pfmFileTypeFile;
      openAttribs.attribs.fileId = helloFileId;
      openAttribs.attribs.fileSize = helloDataSize;
   }
   else
   {
      perr = pfmErrorNotFound;
   }

   op->Complete( perr, &openAttribs, 0);
}

void CCALL Volume::List( PfmMarshallerListOp* op, void* formatterUse)
{
   int64_t openId = op->OpenId();
   int perr = 0;
   PfmAttribs attribs = zeroAttribs;

   if(openId != folderOpenId)
   {
      perr = pfmErrorAccessDenied;
   }
   else
   {
      attribs.fileType = pfmFileTypeFile;
      attribs.fileId = helloFileId;
      attribs.fileSize = helloDataSize;
      op->Add( &attribs, helloFileName);
   }

   op->Complete( perr, true/*noMore*/);
}

void CCALL Volume::ListEnd( PfmMarshallerListEndOp* op, void* formatterUse)
{
   op->Complete( pfmErrorSuccess);
}

void CCALL Volume::Read( PfmMarshallerReadOp* op, void* formatterUse)
{
   int64_t openId = op->OpenId();
   uint64_t fileOffset = op->FileOffset();
   void* data = op->Data();
   size_t requestedSize = op->RequestedSize();
   int perr = 0;
   size_t actualSize = 0;

   if(openId != fileOpenId)
   {
      perr = pfmErrorAccessDenied;
   }
   else
   {
      actualSize = requestedSize;
      if (fileOffset > helloDataSize)
      {
         actualSize = 0;
      }
      else if (fileOffset+requestedSize > helloDataSize)
      {
         actualSize = static_cast<size_t>(helloDataSize-fileOffset);
      }
      memcpy( data, helloData, actualSize);
   }

   op->Complete( perr, actualSize);
}

void CCALL Volume::Write( PfmMarshallerWriteOp* op, void* formatterUse)
{
   op->Complete( pfmErrorAccessDenied, 0/*actualSize*/);
}

void CCALL Volume::SetSize( PfmMarshallerSetSizeOp* op, void* formatterUse)
{
   op->Complete( pfmErrorAccessDenied);
}

void CCALL Volume::Capacity( PfmMarshallerCapacityOp* op, void* formatterUse)
{
   op->Complete( pfmErrorSuccess, helloDataSize/*totalCapacity*/, 0/*availableCapacity*/);
}

void CCALL Volume::FlushMedia( PfmMarshallerFlushMediaOp* op, void* formatterUse)
{
   op->Complete( pfmErrorSuccess, -1/*msecFlushDelay*/);
}

void CCALL Volume::Control( PfmMarshallerControlOp* op, void* formatterUse)
{
   op->Complete( pfmErrorInvalid, 0/*outputSize*/);
}

void CCALL Volume::MediaInfo( PfmMarshallerMediaInfoOp* op, void* formatterUse)
{
   PfmMediaInfo mediaInfo = zeroMediaInfo;

   op->Complete( pfmErrorSuccess, &mediaInfo, L"hellofs"/*mediaLabel*/);
}

void CCALL Volume::Access( PfmMarshallerAccessOp* op, void* formatterUse)
{
   int64_t openId = op->OpenId();
   int perr = 0;
   PfmOpenAttribs openAttribs = zeroOpenAttribs;

   if(openId == folderOpenId)
   {
      openAttribs.openId = folderOpenId;
      openAttribs.openSequence = 1;
      openAttribs.accessLevel = pfmAccessLevelReadData;
      openAttribs.attribs.fileType = pfmFileTypeFolder;
      openAttribs.attribs.fileId = helloRootId;
      perr = 0;
   }
   else if(openId == fileOpenId)
   {
      openAttribs.openId = fileOpenId;
      openAttribs.openSequence = 1;
      openAttribs.accessLevel = pfmAccessLevelReadData;
      openAttribs.attribs.fileType = pfmFileTypeFile;
      openAttribs.attribs.fileId = helloFileId;
      openAttribs.attribs.fileSize = helloDataSize;
      perr = 0;
   }
   else
   {
      perr = pfmErrorNotFound;
   }

   op->Complete( perr, &openAttribs, 0);
}

void CCALL Volume::ReadXattr( PfmMarshallerReadXattrOp* op, void* formatterUse)
{
   op->Complete( pfmErrorNotFound, 0/*xattrSize*/, 0/*transferredSize*/);
}

void CCALL Volume::WriteXattr( PfmMarshallerWriteXattrOp* op, void* formatterUse)
{
   op->Complete( pfmErrorAccessDenied, 0/*transferredSize*/);
}

Volume::Volume(void)
{
   marshaller = 0;
   folderOpenId = 0;
   fileOpenId = 0;
}

Volume::~Volume(void)
{
   ASSERT(!marshaller);
}

int wmain(int argc,const wchar_t*const* argv)
{
   int err = 0;
   PfmApi* pfm = 0;
   PfmMount* mount = 0;
   FD_T toFormatterRead = FD_INVALID;
   FD_T fromFormatterWrite = FD_INVALID;
   PfmMountCreateParams mcp;
   PfmMarshallerServeParams msp;
   Volume volume;

   mcp.toFormatterWrite = FD_INVALID;
   mcp.fromFormatterRead = FD_INVALID;
   msp.dispatch = &volume;
   msp.formatterName = "hellofs";
   msp.toFormatterRead = FD_INVALID;
   msp.fromFormatterWrite = FD_INVALID;

   if (argc <= 1)
   {
      printf(
         "Sample file system application.\n"
         "syntax: hellofs <mount name>\n");
      err = -1;
   }
   else
   {
      mcp.mountSourceName = argv[1];
   }
   if (!err)
   {
      err = PfmApiFactory( &pfm);
      if (err)
      {
         printf( "ERROR: %i Unable to open PFM Api.\n", err);
      }
   }

   if (!err)
   {
      err = PfmMarshallerFactory( &volume.marshaller);
      if (err)
      {
         printf( "ERROR: %i Unable to create marshaller.\n", err);
      }
   }
   if (!err)
   {
         // Communication between the driver and file system is done
         // over a pair of simple pipes. Application is responsible
         // for creating the pipes.
      err = create_pipe( &msp.toFormatterRead, &mcp.toFormatterWrite);
      if (!err)
      {
         err = create_pipe( &mcp.fromFormatterRead, &msp.fromFormatterWrite);
      }
   }
   if (!err)
   {
         // Various mount options are available through mountFlags,
         // visibleProcessId, and ownerSid.
      err = pfm->MountCreate( &mcp, &mount);
      if (err)
      {
         printf( "ERROR: %i Unable to create mount.\n", err);
      }
   }
      // Close driver end pipe handles now. Driver has duplicated what
      // it needs. If these handles are not closed then pipes will not
      // break if driver disconnects, leaving us stuck in the
      // marshaller.
   close_fd( mcp.toFormatterWrite);
   close_fd( mcp.fromFormatterRead);

   if (!err)
   {
         // If tracemon is installed and running then diagnostic
         // messsages can be viewed in the "hellofs" channel.
      volume.marshaller->SetTrace( L"hellofs");
         // Also send diagnostic messages to stdout. This can slow
         // things down quite a bit.
      volume.marshaller->SetStatus( stdout_fd());

         // The marshaller uses alertable I/O, so process can be
         // exited via ctrl+c.
      printf( "Press CTRL+C to exit.\n");
         // The marshaller serve function will return at unmount or
         // if driver disconnects.
      volume.marshaller->ServeDispatch( &msp);
   }

   close_fd( msp.toFormatterRead);
   close_fd( msp.fromFormatterWrite);
   if (mount)
   {
      mount->Release();
   }
   if (pfm)
   {
      pfm->Release();
   }
   if (volume.marshaller)
   {
      volume.marshaller->Release();
      volume.marshaller = 0;
   }
   PfmApiUnload();
   return err;
}
